File: app/src/main/java/com/nicholasworkshop/tinklabstest/fragment/guide/GuideFragment.java
package com.nicholasworkshop.tinklabstest.fragment.guide;

import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.nicholasworkshop.tinklabstest.MainApplication;
import com.nicholasworkshop.tinklabstest.MainComponent;

import javax.inject.Inject;

/**
 * Created by nicholas.wong on 2017/05/15.
 */

public class GuideFragment extends Fragment {

    private static final int INITIAL_STORY_COUNT = 4;
    private static final int FETCH_STORY_COUNT = 10;

    @Inject GuideView guideView;
    @Inject GuideModel guideModel;

    public static GuideFragment newInstance() {
        return new GuideFragment();
    }

    public GuideFragment() {
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        MainComponent component = ((MainApplication) this.getActivity().getApplication()).getComponent();
        component.inject(this);
    }

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, Bundle savedInstanceState) {
        return guideView.createView(inflater, container);
    }

    @Override
    public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        guideModel
                .storyList(INITIAL_STORY_COUNT)
                .subscribe(stories -> guideView.setStoryList(stories));
        guideView
                .storyRequests()
                .flatMap(firstVisibleItemPosition -> guideModel.storyList(firstVisibleItemPosition + FETCH_STORY_COUNT))
                .subscribe(stories -> guideView.setStoryList(stories));
        guideModel
                .ads()
                .subscribe(ads -> guideView.setAdList(ads));
    }
}

File: app/src/main/kotlin/com/nicholasworkshop/tinklabstest/fragment/guide/GuideFragment.kt
package com.nicholasworkshop.tinklabstest.fragment.guide

import android.os.Bundle
import android.support.v4.app.Fragment
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.nicholasworkshop.tinklabstest.MainApplication
import javax.inject.Inject

private const val ARG_GUIDE_TYPE = "ARG_GUIDE_TYPE"

enum class GuideType {
    CITY, SHOP, EAT
}

class GuideFragment : Fragment() {

    companion object {
        fun newInstance(guideType: GuideType): GuideFragment {
            val fragment = GuideFragment()
            val args = Bundle()
            args.putInt(ARG_GUIDE_TYPE, guideType.ordinal)
            fragment.arguments = args
            return fragment
        }

    }

    @Inject
    internal lateinit var guidePresenter: GuidePresenter

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        (this.activity!!.application as MainApplication).component.inject(this)
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View {
        return guidePresenter.createView(inflater, container!!)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        val guideType = GuideType.values()[arguments!!.getInt(ARG_GUIDE_TYPE)]
        guidePresenter.viewCreated(guideType)
    }
}